### ISSUE TITLE
Another issue with subclassing of Any

### ISSUE THREAD

**Issue by mikeshardmind** [microsoft/pyright#7395]
**Describe the bug**

While getting examples together for the intersection PEP, I ran into a few situations that pyright appears to be doing the wrong thing for. 

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFCAKYBuFAhgDYD68CFAUGwMb20DOvUABoAuNlHFQAJhWBRgYMAApeFesACUUALQA%2BTChiiJx8hRgBXECigBGDtz4DBAMSQheMRYLJE46oxIIjhyUNAzMiBReru6e6gB08krqmlAAxPowbC5uHooJSYoARBQg4O5QtCiSULwAFmDm9JJF6myhdEwsUTmx%2BfEARrQg%2BakZvmQD5vil5QK0%2BCDmBlgUZKgwFNUUkvH2PPyExDF5E0L%2BYoHB7dSdEayKvidxiQqj4hkbZPWNzVADFEymFkVRqYBQ9DgwKgCDgICQaDq%2BCQ8345ggqAwvignCqAHJ8FQkGAeJshDcwl1Io9jrkXkkxrUGk0agCoABhBg8Ab0CgAbXigp8xAAuusBPltHoNmwnnT%2BoUij8WSgCf9AQNwABrLbi1oUu7dGlwZ79IYjFIfI6kf7TKCzXDzRbLGCrdYGLbSSRAA)

```python
from typing import Any, reveal_type

class X:
    def foo(self) -> int:
        return 1

class XFirst(X, Any):
    pass

reveal_type(XFirst().foo())  # int
XFirst().foo("errors and should")
reveal_type(XFirst().bar())  # Any, but errors at runtime, intended.

class AnyFirst(Any, X):
    pass

reveal_type(AnyFirst().foo())  # int, should be int if and only if pyright is assuming Any must be compatible with X
reveal_type(AnyFirst().foo)  # should be Callable[..., Any], is () -> int 
AnyFirst().foo("shouldn't be broken, is")
reveal_type(AnyFirst().bar())  # Any, but errors at runtime, intended
```

Given the definition of Any, the prior discussions during the ongoing intersection work, and the behavior which was accepted for subclassing Any (both due to untyped code, and [specifically accepted for `unittest.Mock` like interfaces](https://github.com/python/cpython/issues/91154)) I mostly need to figure out if these three lines are correct for the *current* typing interpretation, if so, many of the arguments about intersections need to be revisited again.

With `AnyFirst().foo()`, it's reasonable to determine int as a minimum bound, but only if Any is assumed to specifically not violate LSP, something I don't think was specified about it, and if it was, I believe that significantly changes the discussion about intersections.

With `AnyFirst.foo` however, this appears completely incorrect. Even assuming there hasn't been an LSP violation, I can come up with multiple user-expressible substitutes for Any which do not cause an LSP violation, yet have a different signature.

```py
class CompatExampleOne:
    def foo(self, *args: Any, **kwargs: Any) -> int: ...

class CompatExampleTwo:
    @overload
    def foo(self) -> int:  ...

    @overload
    def foo(self, *args: Any, **kwargs: Any) -> Any:  ... 
```

subsequently, `AnyFirst().foo("shouldn't be broken, is")` should also be allowed without error.

Narrowing this case specifically, assuming no LSP violation has occured: `AnyFirst().foo()` is the one I'm most concerned about impact of. It could be a reasonable choice to assume no LSP violation has occurred, but I don't believe this matches the specified behavior of Any, and if it does, then we need to go back to the drawing board on intersections, as `Any & T` then becomes "`T` and all of it's possible subtypes" (similar to Kotlin's assumptions, but also something that I was under the impression we determined was not the case)

CC: @erictraut specifically on this one due to impact on prior discussion toward future features, but this may need to go over to discourse either way, as at least one other type checker exhibits the same behavior currently (mypy)

**VS Code extension or command-line**
playground link in this case

**Comment by erictraut** [comment]
Pyright's currently behavior falls out of the MRO linearization it computes in the presence of `Any`. Since the typing spec doesn't currently define how MRO linearization should be computed in the presence of `Any`, I don't think pyright's behavior violates the spec.

If the community decides that this it is sufficiently important for the spec to define how MRO linearization works in the presence of `Any`, then I will update pyright accordingly. Until then, I don't plan to make any changes.

**Comment by mikeshardmind** [comment]
The specification states:

> Any can also be used as a base class. This can be useful for avoiding type checker errors with classes that can duck type anywhere or are highly dynamic.

To me this would indicate that it is highly dynamic from the point where it exists in the MRO as a base class, and this matches the behavior when Any is a singular ancestor, see: https://mail.python.org/archives/list/typing-sig@python.org/thread/GULRKYI7XOB3FLAEFC6OYSTBS5FIA5PU/

However, the actual current behavior is what I'd prefer out of the type system, and I'd want to make intersections consistent with that....

I'll have to open up a discussion on discourse, the answer to this is something that's needed for consistency with or without intersections, but many of the discussions about intersections would have been preempted if pyright's current interpretation is what is deemed correct for the type system.
